packages/arb-token-bridge-ui/src/hooks/useCompleteMultiChainTransactions.ts

```diff
@@ -0,0 +1,331 @@
+  // }
+  // {
+  //   parentChain: ChainId.ArbitrumSepolia,
+  //   chain: ChainId.StylusTestnet
+  // }
+]
+
+function isWithdrawalFromSubgraph(
+  tx: Withdrawal
+): tx is FetchWithdrawalsFromSubgraphResult & AdditionalProperties {
+  return tx.source === 'subgraph'
+}
+
+function isDeposit(tx: DepositOrWithdrawal): tx is Deposit {
+  return tx.direction === 'deposit'
+}
+
+async function transformTransaction(tx: DepositOrWithdrawal) {
+  const parentChainProvider = getProvider(tx.parentChainId)
+  const chainProvider = getProvider(tx.chainId)
+
+  if (isDeposit(tx)) {
+    return transformDeposit(
+      await updateAdditionalDepositData({
+        depositTx: tx,
+        l1Provider: parentChainProvider,
+        l2Provider: chainProvider
+      })
+    )
+  }
+
+  let withdrawal: L2ToL1EventResultPlus | undefined
+
+  if (isWithdrawalFromSubgraph(tx)) {
+    withdrawal = await mapWithdrawalToL2ToL1EventResult({
+      withdrawal: tx,
+      l1Provider: parentChainProvider,
+      l2Provider: chainProvider
+    })
+  } else {
+    if (isTokenWithdrawal(tx)) {
+      withdrawal = await mapTokenWithdrawalFromEventLogsToL2ToL1EventResult({
+        result: tx,
+        l1Provider: parentChainProvider,
+        l2Provider: chainProvider
+      })
+    } else {
+      withdrawal = await mapETHWithdrawalToL2ToL1EventResult({
+        event: tx,
+        l1Provider: parentChainProvider,
+        l2Provider: chainProvider
+      })
+    }
+  }
+
+  if (withdrawal) {
+    return transformWithdrawal(withdrawal)
+  }
+}
+
+function getProvider(chainId: ChainId) {
+  const rpcUrl =
+    rpcURLs[chainId] ?? getWagmiChain(chainId).rpcUrls.default.http[0]
+  return new StaticJsonRpcProvider(rpcUrl)
+}
+
+/**
+ * Fetches transaction history only, and only for a specific direction.
+ * The query could be e.g. deposits or withdrawals.
+ */
+const useTransactionListByDirection = (
+  direction: 'deposits' | 'withdrawals'
+) => {
+  const [transactions, setTransactions] = useState<DepositOrWithdrawal[][]>([])
+  const [page, setPage] = useState(0)
+  const [loading, setLoading] = useState(true)
+
+  const latestTransactions = useLatest(transactions)
+  const latestPage = useLatest(page)
+
+  const { address } = useAccount()
+
+  const shouldFetchPage = useCallback(
+    (index: number) => {
+      const _page = latestPage.current
+
+      if (_page === 0) {
+        return true
+      }
+
+      const txCount = latestTransactions.current[index]?.length ?? 0
+      // fetch next page for a chain pair if all fetched pages are full
+      // we check for >= in case some txs were included by initiating a transfer
+      return txCount / _page >= PAGE_SIZE
+    },
+    [latestTransactions, latestPage]
+  )
+
+  const { data, error } = useSWRImmutable(
+    address ? [direction, address, page] : null,
+    ([_direction, _address, _page]) => {
+      return Promise.all(
+        multiChainFetchList.map((c, index) => {
+          if (!shouldFetchPage(index)) {
```

Comment:
Maybe I didn't understand it right - but shouldn't it be `shouldFetchPage(_page)` instead of `shouldFetchPage(index)` here since we mean to ask whether should if fetch the new page-number (this seems like it's asking for should it fetch the current chain-pair since `index` is tied to that and not `page number`)

Reply:
This method checks if we should fetch next page for the current chain-pair. I see how it could be confusing, the naming isn't nice. I will rename to `shouldFetchNextPage`, I think that should make things clearer?

Reply:
As per our conversation, renaming:

```
shouldFetchPage -> shouldFetchNextPageForChainPair
index -> chainPairIndex
txCount -> txCountForChainPair
```